docs/architecture/adr-020-protobuf-transaction-encoding.md

```diff
@@ -24,76 +25,186 @@ With this in mind, we will tackle the migration path via two main areas, txs and
+encoded body bytes exactly match the signed body bytes with no need for
+["canonicalization"](https://github.com/regen-network/canonical-proto3) of that
+message.
+
+```proto
+// types/types.proto
+message SignDocRaw {
+    bytes body_bytes = 1;
+    string chain_id = 2;
+    uint64 account_number = 3;
+    uint64 account_sequence = 4;
 }
 ```
 
-Note, the `Transaction` type includes `StdTxBase` which will be defined by the SDK
-and includes all the core field members that are common across all transaction types.
-Developers do not have to include `StdTxBase` if they wish, so it is meant to be
-used as an auxiliary type.
+The only requirements are that the client _must_ encode `SignDocRaw` itself canonically.
+ This means that:
+ 
+* all of the fields must be encoded in order
+* default values (i.e. empty/zero values) must be omitted 
 
-### Signing
+If a protobuf implementation does not by default encode `SignDocRaw` canonically,
+the client _must_ manually encode `SignDocRaw` following the guidelines above.
 
-Signing of a `Transaction` must be canonical across clients and binaries. In order
-to provide canonical representation of a `Transaction` to sign over, clients must
-obey the following rules:
+Again, it does not matter if `TxBody` was encoded canonically or not.
 
-- Encode `SignDoc` (see below) via [Protobuf's canonical JSON encoding](https://developers.google.com/protocol-buffers/docs/proto3#json).
-  - Default must be stripped from the output!
-  - JSON keys adhere to their Proto-defined field names.
-- Generate canonical JSON to sign via the [JSON Canonical Form Spec](https://gibson042.github.io/canonicaljson-spec/).
-  - This spec should be trivial to interpret and implement in any language.
+Note that in order to decode `SignDocRaw` for displaying contents to users,
+the regular `SignDoc` type should be used.
 
-```Protobuf
-// app/codec/codec.proto
+3. Broadcast `TxRaw`
 
-message SignDoc {
-  StdSignDocBase base = 1;
-  repeated Message msgs = 2;
+In order to make sure that the signed body bytes exactly match the encoded body
+bytes, clients should encode and broadcast `TxRaw` with the same body bytes used
+in `SignDocRaw`.
+
+```proto
+// types/types.proto
+message TxRaw {
+    bytes body_bytes = 1;
+    repeated bytes signatures = 2;
 }
 ```
 
+Signature verifiers should verify signatures by decoding `TxRaw` and then
+encoding `SignDocRaw` with the raw body bytes.
+
+The standard `Tx` type (which is byte compatible with `TxRaw`) can be used to
+decode transactions for all other use cases.
+
+#### `SIGN_MODE_LEGACY_AMINO`
+
+In order to support legacy wallets and exchanges, Amino JSON will be emporarily
+supported transaction signing. Once wallets and exchanges have had a
+chance to upgrade to protobuf based signing, this option will be disabled. In
+the meantime, it is foreseen that disabling the current Amino signing would cause
+too much breakage to be feasible.
```

Comment:
I think it is worth mentioning that this is a Comsmos Hub focussed requirement and that chains can disallow this singning mode.

Reply:
Sure